google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -574,31 +568,25 @@ private void processOutstandingAckOperations(
       List<PendingModifyAckDeadline> ackDeadlineExtensions) {
     List<PendingModifyAckDeadline> modifyAckDeadlinesToSend =
         Lists.newArrayList(ackDeadlineExtensions);
-    List<String> acksToSend = new ArrayList<>(pendingAcks.size());
-    synchronized (pendingAcks) {
-      if (!pendingAcks.isEmpty()) {
-        try {
-          acksToSend = new ArrayList<>(pendingAcks);
-          logger.log(Level.FINER, "Sending {0} acks", acksToSend.size());
-        } finally {
-          pendingAcks.clear();
-        }
-      }
-    }
+
+    List<String> acksToSend = new ArrayList<>();
+    pendingAcks.drainTo(acksToSend);
+    logger.log(Level.FINER, "Sending {0} acks", acksToSend.size());
+
     PendingModifyAckDeadline nacksToSend = new PendingModifyAckDeadline(0);
-    synchronized (pendingNacks) {
-      if (!pendingNacks.isEmpty()) {
-        try {
-          for (String ackId : pendingNacks) {
-            nacksToSend.addAckId(ackId);
-          }
-          logger.log(Level.FINER, "Sending {0} nacks", pendingNacks.size());
-        } finally {
-          pendingNacks.clear();
-        }
-        modifyAckDeadlinesToSend.add(nacksToSend);
-      }
+    pendingNacks.drainTo(nacksToSend.ackIds);
+    logger.log(Level.FINER, "Sending {0} nacks", nacksToSend.ackIds.size());
+    if (!nacksToSend.ackIds.isEmpty()) {
+      modifyAckDeadlinesToSend.add(nacksToSend);
     }
+
+    PendingModifyAckDeadline receiptsToSend = new PendingModifyAckDeadline(getMessageDeadlineSeconds());
+    pendingReceipts.drainTo(receiptsToSend.ackIds);
+    logger.log(Level.FINER, "Sending {0} receipts", receiptsToSend.ackIds.size());
+    if (!receiptsToSend.ackIds.isEmpty()) {
+      modifyAckDeadlinesToSend.add(receiptsToSend);
```

Comment:
Max, you said that it's OK to ack and modack the same ID in one request. Is it also OK to modack the same ID twice?

It's possible that we nack quickly, and get modacks for both nack and receipt.

Reply:
That should be alright as well, I'll double check but I agree that simplifying the batching is more important than guaranteeing uniqueness.